design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,148 @@
+  - _Note:_ the `provisioner` field is ignored.  Defining the provisioner is unnecessary as Rook-Ceph is the only operator watching for CephObjectBucketClaims.
+
+**Use Case: Provision a Bucket** 
+
+_As a Kubernetes user, I want to leverage the Kubernetes API to create Ceph Object Buckets. I expect to get back the bucket connection information in a Pod-attachable object._
+ 
+![user-actions](./cobc-user.png)
+
+0. The user creates a CephObjectBucketClaim in the user's namespace.
+1. The Rook Operator detects a new ObjectBucketClaim instance.  
+    1. The operator gets the user's credentials via `cephObjectBucketClaim.spec.secretName`.  
+    1. The operator gets the object store URL via `storageClass.parameters.serviceNamespace` + `storageClass.parameters.serviceName`.
+1. The operator uses the object store URL and credentials to make an S3 "make bucket" call.
+1. The operator creates a ConfigMap in the namespace of the CephObjectBucketClaim with relevant connection data for that bucket.
+1. An app Pod may then mount the Secret and the ConfigMap to begin accessing the bucket. 
+
+#####Use Case: Delete an Object Bucket
+
+_As a Kubernetes user, I want to delete ObjectBucketClaim instances and cleanup generated API resources._
+
+1. The user deletes the CephObjectBucketClaim via `kubectl delete ...`.
+1. The CephObjectBucketClaim is marked for deletion and left in the foreground.
+1. The respective ConfigMap is deleted.
+1. The CephObjectBucketClaim is garbage collected.
+
+---
+
+## API Specifications
+
+### Custom Resource Definition
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1 
+kind: CustomResourceDefinition
+metadata:
+  name: cephobjectbucketclaims.rook.io
+spec:
+  group: ceph.rook.io
+  version: v1
+  scope: namespaced
+  names:
+      kind: CephObjectBucketClaim
+      listKind: CephObjectBucketClaimList
+      singular: cephobjectbucketclaim
+      plural: cephobjectbucketclaims
+```
+
+### Custom Resource
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephObjectBucketClaim
+metadata:
+  name: my-bucket-1
+  namespace: dev-user
+  labels:
+    rook.io/bucket-provisioner:
+    rook.io/object-bucket-claim: my-bucket-1 [1]
+spec:
+  storageClassName: some-object-store
+  secretName: my-s3-key-pair
+  bucketNamePrefix: prefix [2]
+```
+
+1. Added by the Rook-Ceph operator.
+1. The operator will append a hyphen followed by random characters to the string given here.
+
+### ConfigMap
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: rook-ceph-object-bucket-my-bucket-1
+  namespace: dev-user
+  labels:
+    ceph.rook.io/claims:
+  ownerReferences:  [1]
+  - name: my-bucket-1
+    uid: 1234-qwer-4321-rewq
+    apiVersion: ceph.rook.io/v1
+    kind: CephObjectBucketClaim
+    blockOwnerDeletion: true 
+data:
+  CEPH_BUCKET_HOST: http://my-store-url-or-ip
+  CEPH_BUCKET_NAME: my-bucket-1
+  CEPH_BUCKET_PORT: 80
+  CEPH_BUCKET_SSL: no
+```
```

Comment:
do we want a finalizer to prevent the CM from being deleted separate from the bucket?

Reply:
From the previous PR, if I'm remembering it accurately, the consensus was to avoid having to handle finalizers ourselves.

Reply:
I recall some discussion like that from @travisn I think. I believe the concern is there is an extra burden (manually deleting the finalizer) placed on the user who has a reason for deleting the CM.

Reply:
Right, I'd suggest using finalizers sparingly and only when we have a significant need to protect a resource from being deleted. In this case, if the user did unintentionally delete the CM, the operator could regenerate it and no loss of data, right?

Reply:
I think that's a reasonable solution.  Documentation can explain that if a user wants to permanently remove the CM, they should delete the claim.

Reply:
Agree with using finalizers sparingly. The thinking here is that the OB and OBC controllers need to reconcile current state with expected state.  If an OBC is deleted the CM and secret should be deleted (which is automated by our use of `ownerReference`. If an OB is deleted the OBC needs to be updated to reflect it it no longer bound. If the secret or CM are deleted we have an OBC that can no longer reach the OB -- part of the binding has broken. We can either watch secrets and CMs, or we can slap a finalizer on the generated secret and CM to reduce the chances of the CM or secret being accidentally deleted. They can still be deleted but it is more deliberate now since the user must first delete the finalizer.